### PR TITLE
Provisioning: Allow startup when there are dashboard provisioning failures

### PR DESCRIPTION
**Pull request by suntala**
**What is this feature?**

Log dashboard provisioning issues instead of preventing startup.

**Why do we need this feature?**

Issues restarting Grafana because a folder with provisioned dashboards was deleted and then recreated via UI.

**Who is this feature for?**

Users who rely on provisioning

**Which issue(s) does this PR fix?**:

Relates to https://github.com/grafana/support-escalations/issues/10964

**Special notes for your reviewer:**

* Need to read about backport policies but thinking we should backport to the same versions that https://github.com/grafana/grafana/pull/87465 does: v9.5.x, v10.4.x and v11.0.x. This was a fix for a similar issue.
* See doc [[Azure escalation] Provisioning when the target folder is deleted](https://docs.google.com/document/d/190kiLZVl4_0CI4MNR4M0nqUkIslK2W85ARvS39CozXw/edit). This PR implements proposal 6.
* We're going to need frontend changes for this as well: "I wonder whether we could use something similar to the alerts for license expiration to throw up a visible warning inside grafana itself for admins" ([source](https://raintank-corp.slack.com/archives/C05FYAPEPKP/p1719415987229529?thread_ts=1719401988.624259&cid=C05FYAPEPKP)). IMO this could be done in a separate PR. We can prioritise alleviating the current bug and then make the change clearer in the UI. --> created ticket: https://github.com/grafana/grafana/issues/92280
* This PR is not instead of https://github.com/grafana/grafana/pull/92022 but in addition to.
* Dashboards aren't the only question when it comes to provisioning. Alert rules can also be provisioned to specific folders. So far I haven't found any reference to folder UIDs when it comes to alert rule provisioning just folder names (example [doc](https://grafana.com/docs/grafana/latest/alerting/set-up/provision-alerting-resources/file-provisioning/) I looked at). Therefore, we shouldn't be running into similar issues when it comes to them. I'll double check with the alerting team though.

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

File: pkg/services/provisioning/provisioning.go [hunk 1]
```diff
@@ -185,7 +185,7 @@ func (ps *ProvisioningServiceImpl) Run(ctx context.Context) error {
 	err := ps.ProvisionDashboards(ctx)
 	if err != nil {
 		ps.log.Error("Failed to provision dashboard", "error", err)
-		return err
+		return nil
```
Comment (DanCech) [inline pkg/services/provisioning/provisioning.go:188]: I am wondering whether we should actually remove this line and allow the logic to continue and start the dashboard poller. That would at least give us a chance at provisioning the dashboard correctly if the problem gets resolved, and will continue to complain in the logs until the problem is fixed, which might make it easier for an admin to figure out what's going on vs having to find a single log line at startup.

Comment (suntala) [reply pkg/services/provisioning/provisioning.go]: That makes a lot of sense. What's nice is that in the case we've seen with Azure, as soon as the user deletes the duplicate folder (the one with the wrong UID), the right one gets recreated automatically by the provisioning service without needing to restart Grafana.

**Review by owensmallwood — APPROVED** [review]
LGTM 🚀

**Comment by dreddy-work** [comment]
Hi @suntala, Could you help us understand what’s preventing the backport here? The change in this PR seems fairly small (excluding tests). I noticed you also have PRs opened directly to the backporting branches, but I'm not sure if those are meant to address this issue. A backport would be really helpful for us, as most of our customers are on Grafana 10.*.

**Comment by dreddy-work** [comment]
Thank you @suntala for the update. One of my colleagues found a variant of this issue and failure seems coming from this line: https://github.com/grafana/grafana/blame/733ae1f0996b8475456dab363a07089127edd58c/pkg/services/provisioning/dashboards/file_reader.go#L382.  Seems pre provisioned dashboard **folder** is deleted and then created a new dashboard with same name (instead of folder) at the same root level causing the provision to fail. Do you expect this issue also be fixed with this change?

**Comment by suntala** [comment]
@dreddy-work thank you for the clear reproduction steps. As you mention, pre-change this would have prevented Grafana from starting up.

https://github.com/grafana/grafana/pull/92560 (the change) means that the variant of the issue will not stop Grafana from starting up. However, users will still need to delete/rename the new dashboard (the one that created the naming conflict). This is so that the folder which has been designated to contain provisioned dashboards can be created. I hope that answers your question and please let me know if you have any others.

**Pull request merged**